File: Go/ExampleTests/common/shoppingcart_string.go
package common

type ShoppingCartString struct {
	Items string
	Shipping string
	Discount string
	TotalPrice string
	ShippingAddress AddressString
	BillingAddress AddressString
}

func NewShoppingCartStringFromSlice(v []string) ShoppingCartString {
	s := ShoppingCartString{}
	if len(v) > 0 { s.Items = v[0] }
	if len(v) > 1 { s.Shipping = v[1] }
	if len(v) > 2 { s.Discount = v[2] }
	if len(v) > 3 { s.TotalPrice = v[3] }
	if len(v) > 4 { s.ShippingAddress = NewAddressStringFromText(v[4]) }
	if len(v) > 5 { s.BillingAddress = NewAddressStringFromText(v[5]) }
	return s
}

// NewShoppingCartStringFromText builds from the text form, e.g. Money as "25 USD".
func NewShoppingCartStringFromText(text string) ShoppingCartString {
	parts := RequireTokens(text, 6, "ShoppingCart")
	return ShoppingCartString{
		Items: parts[0],
		Shipping: parts[1],
		Discount: parts[2],
		TotalPrice: parts[3],
		ShippingAddress: NewAddressStringFromText(parts[4]),
		BillingAddress: NewAddressStringFromText(parts[5]),
	}
}

func (s ShoppingCartString) String() string {
	return Token(s.Items) + " " + Token(s.Shipping) + " " + Token(s.Discount) + " " + Token(s.TotalPrice) + " " + Nested(s.ShippingAddress.String()) + " " + Nested(s.BillingAddress.String())
}

func (s ShoppingCartString) Equals(o ShoppingCartString) bool {
	return DNCEqual(s.Items, o.Items) &&
		DNCEqual(s.Shipping, o.Shipping) &&
		DNCEqual(s.Discount, o.Discount) &&
		DNCEqual(s.TotalPrice, o.TotalPrice) &&
		s.ShippingAddress.Equals(o.ShippingAddress) &&
		s.BillingAddress.Equals(o.BillingAddress)
}

func EqualShoppingCartStringSlices(a, b []ShoppingCartString) bool {
	if len(a) != len(b) { return false }
	for i := range a {
		if !a[i].Equals(b[i]) { return false }
	}
	return true
}

File: Rust/ExampleTests/common/shoppingcart_string.rs
#![allow(dead_code, unused_imports, unused_variables)]

use super::address_string::AddressString;
#[derive(Debug, Clone, Default)]
pub struct ShoppingCartString {
    pub items: String,
    pub shipping: String,
    pub discount: String,
    pub totalprice: String,
    pub shippingaddress: AddressString,
    pub billingaddress: AddressString,
}

impl ShoppingCartString {
    pub fn from_vec(v: &[&str]) -> Self {
        Self {
            items: v.get(0).copied().unwrap_or("").to_string(),
            shipping: v.get(1).copied().unwrap_or("").to_string(),
            discount: v.get(2).copied().unwrap_or("").to_string(),
            totalprice: v.get(3).copied().unwrap_or("").to_string(),
            shippingaddress: Default::default(),
            billingaddress: Default::default(),
        }
    }

    /// Builds from the text form, e.g. Money as "25 USD".
    pub fn from_text(text: &str) -> Self {
        let parts = crate::common::tokens::require(text, 6, "ShoppingCart");
        Self {
            items: parts[0].clone(),
            shipping: parts[1].clone(),
            discount: parts[2].clone(),
            totalprice: parts[3].clone(),
            shippingaddress: AddressString::from_text(&parts[4]),
            billingaddress: AddressString::from_text(&parts[5]),
        }
    }
}

impl std::fmt::Display for ShoppingCartString {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", [crate::common::tokens::token(&self.items), crate::common::tokens::token(&self.shipping), crate::common::tokens::token(&self.discount), crate::common::tokens::token(&self.totalprice), crate::common::tokens::nested(&self.shippingaddress.to_string()), crate::common::tokens::nested(&self.billingaddress.to_string())].join(" "))
    }
}

impl PartialEq for ShoppingCartString {
    fn eq(&self, other: &Self) -> bool {
        crate::common::dnc_equal(&self.items, &other.items)
            && crate::common::dnc_equal(&self.shipping, &other.shipping)
            && crate::common::dnc_equal(&self.discount, &other.discount)
            && crate::common::dnc_equal(&self.totalprice, &other.totalprice)
            && self.shippingaddress == other.shippingaddress
            && self.billingaddress == other.billingaddress
    }
}
